File: pine-java/examples/MultiPipelineServer.java
/*
 * Multi-Pipeline Server — Pine-Java Embedding Example
 * ===================================================
 *
 * One process, several pipelines, each bound to its own endpoint with its
 * own log prefix:
 *
 *   POST /api/feed    -> feed.json    (log_prefix "[feed] ")
 *   POST /api/search  -> search.json  (log_prefix "[search] ")
 *   POST /execute     -> 410 Gone     (legacy endpoint, deliberately retired)
 *
 * Instead of running the bundled single-pipeline PineServer (whose start()
 * owns /execute), each pipeline is an EMBEDDED runtime: PineServer.load()
 * builds the engine + resources + hot-reload watcher without any HTTP, and
 * PineServer.execute(common, items) runs a request against the live
 * reference-counted snapshot (issue #169 embedding API). The HTTP layer below
 * is a plain com.sun.net.httpserver mux the application owns — the same
 * pattern works under Spring/Vert.x by calling execute() from your handlers.
 *
 * Since issue #172 log_prefix is engine-scoped: lines emitted while the feed
 * pipeline executes (observe_log, [pine-debug]) carry "[feed] " while search
 * lines carry "[search] ", concurrently, in one process. No global state.
 *
 * /execute is kept only as a tombstone: it answers 410 Gone and points
 * callers at the named endpoints.
 *
 * Compile & run (from repo root; the default Maven build compiles this
 * example as a test source, so plain `mvn test-compile` catches rot):
 *
 *   cd pine-java && mvn -q test-compile
 *   CP="target/classes:target/test-classes:$(mvn -q dependency:build-classpath -Dmdep.outputFile=/dev/stdout | tail -1)"
 *   java -cp "$CP" page.liam.pine.examples.MultiPipelineServer \
 *        ../pine-go/examples/multi-pipeline/feed.json \
 *        ../pine-go/examples/multi-pipeline/search.json 8080
 *
 * Try:
 *
 *   curl -s -X POST localhost:8080/api/feed   -d '{"common":{"user_id":"u1"}}'
 *   curl -s -X POST localhost:8080/api/search -d '{"common":{"query":"tech"}}'
 *   curl -si -X POST localhost:8080/execute   -d '{}'   # 410 Gone
 */
package page.liam.pine.examples;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import page.liam.pine.Engine;
import page.liam.pine.PineErrors;
import page.liam.pine.PineServer;

import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class MultiPipelineServer {
    private static final ObjectMapper mapper = new ObjectMapper();

    public static void main(String[] args) throws Exception {
        String feedConfig = args.length > 0 ? args[0] : "feed.json";
        String searchConfig = args.length > 1 ? args[1] : "search.json";
        int port = args.length > 2 ? Integer.parseInt(args[2]) : 8080;

        // One embedded runtime per pipeline. load() builds the engine,
        // starts resources and the config hot-reload watcher — no HTTP.
        // Each config declares its own log_prefix; since issue #172 the
        // prefix scopes to its engine, so both runtimes below log under
        // different prefixes in the same process.
        PineServer feed = new PineServer(feedConfig, 0);
        feed.load();
        PineServer search = new PineServer(searchConfig, 0);
        search.load();

        HttpServer http = HttpServer.create(new InetSocketAddress(port), 0);
        // HttpServer contexts use longest-PREFIX matching: without a guard,
        // /api/feed/anything would run the feed pipeline. exact() rejects
        // sub-paths with 404, matching the bundled PineServer.wrapHandler
        // and the Go/C++ examples' exact-path dispatch.
        http.createContext("/api/feed", exact("/api/feed", exchange -> handlePipeline(exchange, feed)));
        http.createContext("/api/search", exact("/api/search", exchange -> handlePipeline(exchange, search)));
        // Legacy endpoint: kept so old callers get a clear migration signal
        // instead of a generic 404, but it no longer runs any pipeline.
        http.createContext("/execute", exact("/execute", exchange -> respond(exchange, 410,
                Map.of("error", "endpoint retired: use /api/feed or /api/search"))));
        http.start();
        System.out.println("listening on :" + port);

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            http.stop(2);
            feed.stop();
            search.stop();
        }));
    }

    // Mirrors the bundled server's default request-body cap. An embedding
    // HTTP layer must keep this boundary itself — issue #169 treats the body
    // limit as a shared-dispatch-layer safety contract.
    private static final long MAX_BODY_BYTES = 10L * 1024 * 1024; // 10 MB

    // Adapts HTTP to one embedded pipeline runtime. execute() acquires the
    // live snapshot with an in-flight reference, so a concurrent hot-reload
    // never tears the engine down mid-request.
    private static void handlePipeline(HttpExchange exchange, PineServer rt) throws java.io.IOException {
        if (!"POST".equals(exchange.getRequestMethod())) {
            respond(exchange, 405, Map.of("error", "method not allowed"));
            return;
        }
        byte[] body = readLimitedBody(exchange, MAX_BODY_BYTES);
        if (body == null) {
            respond(exchange, 413, Map.of("error", "request body too large"));
            return;
        }
        Map<String, Object> common;
        List<Map<String, Object>> items;
        try {
            Map<String, Object> parsed = mapper.readValue(body, new TypeReference<>() {});
            @SuppressWarnings("unchecked")
            Map<String, Object> c = (Map<String, Object>) parsed.get("common");
            @SuppressWarnings("unchecked")
            List<Map<String, Object>> it = (List<Map<String, Object>>) parsed.get("items");
            common = c != null ? c : Map.of();
            items = it != null ? it : List.of();
        } catch (Exception e) {
            respond(exchange, 400, Map.of("error", "invalid request body"));
            return;
        }

        try {
            Engine.Result result = rt.execute(common, items);
            Map<String, Object> resp = new LinkedHashMap<>();
            resp.put("common", result.common);
            resp.put("items", result.items);
            // Operator failures are RETURNED in result.error, not thrown —
            // the Java throw/return split (validation throws, execution
            // returns). Mirror the bundled handler: error != null -> 500.
            if (result.error != null) {
                if (result.error instanceof PineErrors.PanicError) {
                    // Log the stack trace server-side; clients get the safe message.
                    System.err.println("[example] PANIC: "
                            + ((PineErrors.PanicError) result.error).detailedError());
                }
                resp.put("error", result.error.getMessage());
            }
            respond(exchange, result.error != null ? 500 : 200, resp);
        } catch (PineErrors.ValidationError e) {
            respond(exchange, 400, Map.of("error", e.getMessage()));
        } catch (Exception e) {
            respond(exchange, 500, Map.of("error", e.getMessage()));
        }
    }

    // Reads at most limit bytes; returns null once the body exceeds it (the
    // caller answers 413). Counting the stream keeps memory bounded without
    // trusting Content-Length.
    private static byte[] readLimitedBody(HttpExchange exchange, long limit) throws java.io.IOException {
        java.io.ByteArrayOutputStream buf = new java.io.ByteArrayOutputStream();
        byte[] chunk = new byte[8192];
        long total = 0;
        java.io.InputStream in = exchange.getRequestBody();
        int n;
        while ((n = in.read(chunk)) != -1) {
            total += n;
            if (total > limit) {
                return null;
            }
            buf.write(chunk, 0, n);
        }
        return buf.toByteArray();
    }

    // Enforces exact path matching over HttpServer's longest-prefix contexts.
    private static com.sun.net.httpserver.HttpHandler exact(
            String path, com.sun.net.httpserver.HttpHandler handler) {
        return exchange -> {
            if (!path.equals(exchange.getRequestURI().getPath())) {
                respond(exchange, 404, Map.of("error", "not found"));
                return;
            }
            handler.handle(exchange);
        };
    }

    private static void respond(HttpExchange exchange, int status, Object body) throws java.io.IOException {
        byte[] bytes = mapper.writeValueAsBytes(body);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, bytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }
}

File: pine-cpp/examples/multi_pipeline_server.cpp
// Multi-pipeline server example — one process, several pipelines, each bound
// to its own endpoint with its own log prefix.
//
// Layout:
//
//   POST /api/feed    -> feed.json    (log_prefix "[feed] ")
//   POST /api/search  -> search.json  (log_prefix "[search] ")
//   POST /execute     -> 410 Gone     (legacy endpoint, deliberately retired)
//
// pine-cpp's bundled Server is not a public library API, so this example
// embeds the ENGINE layer directly (include/pine/pine.hpp): one pine::Engine
// per pipeline, plus a deliberately tiny blocking HTTP loop the application
// owns. The same pattern drops into any C++ HTTP framework (Drogon, Pistache,
// oatpp) by calling engine.execute() from your handlers. Engine::execute is
// concurrency-safe; production servers should also add hot-reload and
// graceful drain (see src/server/server.cpp for the reference shape).
//
// Since issue #172 log_prefix is engine-scoped: lines emitted while the feed
// pipeline executes (observe_log, [pine-debug]) carry "[feed] " while search
// lines carry "[search] ", concurrently, in one process.
//
// /execute is kept only as a tombstone: it answers 410 Gone and points
// callers at the named endpoints.
//
// Build (from pine-cpp/, requires the configured build dir):
//
//   cmake --build build --target multi_pipeline_server -j8
//
// Run & try:
//
//   ./build/multi_pipeline_server ../pine-go/examples/multi-pipeline/feed.json
//       ../pine-go/examples/multi-pipeline/search.json 8080   (one command line)
//   curl -s -X POST localhost:8080/api/feed   -d '{"common":{"user_id":"u1"}}'
//   curl -s -X POST localhost:8080/api/search -d '{"common":{"query":"tech"}}'
//   curl -si -X POST localhost:8080/execute   -d '{}'   # 410 Gone

#include "pine/pine.hpp"

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>

#include <cerrno>
#include <csignal>
#include <cstring>
#include <iostream>
#include <map>
#include <memory>
#include <sstream>
#include <string>

namespace {

pine::Request parse_request(const std::string& body) {
  pine::Request request;
  const auto root = pine::parse_json(body).as_object();
  if (auto it = root.find("common"); it != root.end()) {
    for (const auto& [key, value] : it->second.as_object()) {
      request.common[key] = value;
    }
  }
  if (auto it = root.find("items"); it != root.end()) {
    for (const auto& item : it->second.as_array()) {
      pine::Variant::object_t row;
      for (const auto& [key, value] : item.as_object()) {
        row[key] = value;
      }
      request.items.push_back(std::move(row));
    }
  }
  return request;
}

void send_response(int fd, int status, const std::string& body) {
  const char* text = status == 200   ? "OK"
                     : status == 400 ? "Bad Request"
                     : status == 404 ? "Not Found"
                     : status == 405 ? "Method Not Allowed"
                     : status == 410 ? "Gone"
                                     : "Internal Server Error";
  std::ostringstream resp;
  resp << "HTTP/1.1 " << status << " " << text << "\r\n"
       << "Content-Type: application/json\r\n"
       << "Content-Length: " << body.size() << "\r\n"
       << "Connection: close\r\n\r\n"
       << body;
  const std::string out = resp.str();
  // MSG_NOSIGNAL is mandatory on every pine-cpp raw-socket write path
  // (llmdoc/must/conventions.md): a client that disconnects mid-response
  // would otherwise raise SIGPIPE and kill the whole process. Loop until
  // all bytes are written — a single send may be short.
  size_t sent = 0;
  while (sent < out.size()) {
    ssize_t n = ::send(fd, out.data() + sent, out.size() - sent, MSG_NOSIGNAL);
    if (n < 0 && errno == EINTR) {
      continue;
    }
    if (n <= 0) {
      break;  // peer gone; nothing sensible left to do for this response
    }
    sent += static_cast<size_t>(n);
  }
}

std::string json_error(const std::string& message) {
  // Route the message through the Variant serializer: exception text can
  // contain quotes (ValidationError quotes field names) or control
  // characters, and hand-concatenating it would produce invalid JSON.
  pine::Variant::object_t obj;
  obj["error"] = pine::Variant(message);
  std::string out = pine::dump_json(pine::Variant(std::move(obj)), 0);
  if (out.empty() || out.back() != '\n') {
    out += '\n';
  }
  return out;
}

// Adapts one HTTP request to one embedded engine. execute() is
// concurrency-safe; each engine's diagnostics carry its own log_prefix.
void handle_pipeline(int fd, pine::Engine& engine, const std::string& method, const std::string& body) {
  if (method != "POST") {
    send_response(fd, 405, json_error("method not allowed"));
    return;
  }
  pine::Request request;
  try {
    request = parse_request(body);
  } catch (const std::exception&) {
    send_response(fd, 400, json_error("invalid request body"));
    return;
  }
  try {
    pine::Result result = engine.execute(request);
    send_response(fd, 200, pine::result_to_json(result));
  } catch (const pine::ValidationError& e) {
    send_response(fd, 400, json_error(e.what()));
  } catch (const std::exception& e) {
    send_response(fd, 500, json_error(e.what()));
  }
}

volatile std::sig_atomic_t g_stop = 0;

}  // namespace

int main(int argc, char** argv) {
  const std::string feed_config = argc > 1 ? argv[1] : "feed.json";
  const std::string search_config = argc > 2 ? argv[2] : "search.json";
  const int port = argc > 3 ? std::atoi(argv[3]) : 8080;

  // One embedded engine per pipeline. Each config declares its own
  // log_prefix; since issue #172 the prefix is an engine-instance member,
  // so both engines below log under different prefixes in one process.
  std::unique_ptr<pine::Engine> feed;
  std::unique_ptr<pine::Engine> search;
  try {
    feed = std::make_unique<pine::Engine>(pine::load_config_from_file(feed_config));
    search = std::make_unique<pine::Engine>(pine::load_config_from_file(search_config));
  } catch (const std::exception& e) {
    std::cerr << "error creating engines: " << e.what() << "\n";
    return 1;
  }

  int listen_fd = ::socket(AF_INET, SOCK_STREAM, 0);
  int reuse = 1;
  ::setsockopt(listen_fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse));
  sockaddr_in addr{};
  addr.sin_family = AF_INET;
  addr.sin_addr.s_addr = htonl(INADDR_ANY);
  addr.sin_port = htons(static_cast<uint16_t>(port));
  if (::bind(listen_fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) != 0 ||
      ::listen(listen_fd, 16) != 0) {
    std::cerr << "failed to listen on :" << port << "\n";
    return 1;
  }
  std::signal(SIGINT, [](int) { g_stop = 1; });
  std::cout << "listening on :" << port << "\n";

  // Minimal single-threaded accept loop — enough to demonstrate routing;
  // real deployments should use a proper HTTP framework or the bundled
  // pineapple-server shape (per-connection threads, timeouts, drain).
  while (!g_stop) {
    int fd = ::accept(listen_fd, nullptr, nullptr);
    if (fd < 0) {
      break;
    }
    std::string raw(64 * 1024, '\0');
    ssize_t n = ::read(fd, raw.data(), raw.size());
    if (n <= 0) {
      ::close(fd);
      continue;
    }
    raw.resize(static_cast<size_t>(n));

    std::istringstream head(raw.substr(0, raw.find("\r\n")));
    std::string method, path;
    head >> method >> path;
    const auto body_pos = raw.find("\r\n\r\n");
    const std::string body = body_pos == std::string::npos ? "" : raw.substr(body_pos + 4);

    if (path == "/api/feed") {
      handle_pipeline(fd, *feed, method, body);
    } else if (path == "/api/search") {
      handle_pipeline(fd, *search, method, body);
    } else if (path == "/execute") {
      // Legacy endpoint: kept so old callers get a clear migration signal
      // instead of a generic 404, but it no longer runs any pipeline.
      send_response(fd, 410, json_error("endpoint retired: use /api/feed or /api/search"));
    } else {
      send_response(fd, 404, json_error("not found"));
    }
    ::close(fd);
  }

  ::close(listen_fd);
  feed->close();
  search->close();
  return 0;
}
